File: prototype/autodiff.py
from __future__  import annotations
from collections import defaultdict
from typing import Dict

"""
This method contains reverse-mode
"""

class Symbol: 
    def __init__(self, value:float, local_gradients:tuple = ()) -> None:
        self.value = value 
        self.local_gradients = local_gradients


    def __add__(self, other:Symbol) -> Symbol:
        return Symbol(
            value = self.value + other.value, 
            local_gradients = (
                (self, 1.0),
                (other, 1.0)
            )
        )

    def __sub__(self, other:Symbol) -> Symbol:
        return Symbol(
            value = self.value - other.value, 
            local_gradients = (
                (self, 1.0),
                (other, 1.0)
            )
        )

    def __mul__(self, other:Symbol) -> Symbol:
        return Symbol(
            value = self.value * other.value, 
            local_gradients = (
                (self, other.value),
                (other, self.value)
            )
        )

    def __neg__(self) -> Symbol:
        return Symbol(
            value = -1 * self.value,
            local_gradients=(
                (self, -1)
            )
        )

    def __truediv__(self, other:Symbol) -> Symbol:
        return self * Symbol.inv(other)

    @staticmethod
    def inv(other) -> Symbol:
        return Symbol(
            value = 1.0 / other.value,
            local_gradients = (
                (other, -1/(pow(other.value, 2)))
            )
        )

def gradients(variable:Symbol) -> Dict[Symbol, float]: 
    _gradients:dict = defaultdict(lambda: 0.0)

    def _compute_gradients(variable:Symbol, path_value:float) -> None:
        print(type(variable.local_gradients))

        # for child, local_gradient in variable.local_gradients:
        #     value_of_path_to_child = path_value * local_gradient
        #     _gradients[child] += value_of_path_to_child
        #     _compute_gradients(child, value_of_path_to_child)

    _compute_gradients(variable, path_value = 1)

    return _gradients

if __name__ == '__main__':
    
    def f(a: Symbol, b: Symbol) -> Symbol:
        return (a / b - a) * (b / a + a + b) * (a - b)

    a = Symbol(230.3)
    b = Symbol(33.2)
    y = f(a, b)

    df = gradients(y)

    print(f"d value: {y.value}")
    print(f"df/da: {df[a]}", f"df/db: {df[b]}", sep='\n')

File: prototype/autodiff.cpp
#include <functional>
#include <iostream>
#include <iomanip>
#include <map>

#include <list>
#include <utility>
#include <cmath>

namespace ad {

class Symbol {
private:
  std::list<std::pair<Symbol, double>> m_local_gradient;
  double m_value{};

public:
  Symbol(const std::list<std::pair<Symbol, double>> &_local_gradient, double _value)
      : m_local_gradient(_local_gradient), m_value(_value) {}

  Symbol(double _value) : m_value(_value) {}

  auto value() const noexcept -> double { return m_value; }

  auto local_gradient() const noexcept -> std::list<std::pair<Symbol, double>> {
    return m_local_gradient;
  }

  auto operator<(const Symbol &rhs) const -> bool {
    return m_value < rhs.m_value;
  }

  auto operator>(const Symbol &rhs) const -> bool {
    return m_value > rhs.m_value;
  }

  auto operator==(const Symbol &rhs) const -> bool {
    return m_value == rhs.m_value;
  }

  auto operator!=(const Symbol &rhs) const -> bool {
    return m_value != rhs.m_value;
  }
};

auto operator+(const Symbol &lhs, const Symbol &rhs) -> Symbol {
  return {
    {{lhs, 1.0}, {rhs, 1.0}},
    lhs.value() + rhs.value()
  };
}

auto operator-(const Symbol &lhs, const Symbol &rhs) -> Symbol {
  return {
    {{lhs, 1.0}, {rhs, -1.0}},
    lhs.value() - rhs.value()
  };
}

auto operator*(const Symbol &lhs, const Symbol &rhs) -> Symbol {
  return {
    {{lhs, rhs.value()}, {rhs, lhs.value()}},
    lhs.value() * rhs.value()
  };
}

auto operator/(const Symbol &lhs, const Symbol &rhs) -> Symbol {
  Symbol inverse{
    {{rhs, -1.0 / std::pow(rhs.value(), 2)}},
    1.0 / rhs.value()
  };

  return lhs * inverse;
}

auto gradient(const Symbol &variable) -> std::map<Symbol, double> {
  std::map<Symbol, double> _gradients{};

  std::function<auto(const Symbol &, double) -> void> _compute_gradient =
      [&_compute_gradient, &_gradients](const Symbol &variable,
                                        double path_value) {
        double value_of_path_to_child{};
        std::cout << "outer: " << path_value << '\n';

        for (auto &[child, local_gradient] : variable.local_gradient()) {
          value_of_path_to_child = path_value * local_gradient;
          
          // _gradients.emplace(child, 0);

          // std::cout << std::setprecision(20)
          //           << "\t local_gradient: " << local_gradient << '\n';
          // std::cout << std::setprecision(20)
          //           << "\t path value: " << path_value<< '\n';
          // std::cout << std::setprecision(20)
          //           << "\t gradient before: " << _gradients[child] << '\n';

          _gradients[child] += value_of_path_to_child;
          // if(_gradients.contains(child)){
          //   _gradients.at(child) += value_of_path_to_child;
          // } else {
          //   _gradients[child] = value_of_path_to_child;
          // }

          // std::cout << std::setprecision(20)
          //           << "\t gradient after: " << _gradients[child] << '\n';

          _compute_gradient(child, value_of_path_to_child);
        }
      };

  _compute_gradient(variable, 1);

  return _gradients;
}

} // namespace ad

auto main(void) -> int {
  using namespace ad;

  // auto a = Symbol(4.0f);
  // auto b = Symbol(3.0f);
  // auto c = a + b;
  // auto d = a * c / a;

  // auto df = gradient(d);
  // std::cout << d.value() << '\n';
  // std::cout << df[a] << '\n';
  // std::cout << df[b] << '\n';

  Symbol a{230.3};
  Symbol b{33.2};

  auto y = (a / b - a) * (b / a + a + b) * (a - b);
  auto df = gradient(y);

  std::cout << "d value: " << std::setprecision(20) << y.value() << '\n';
  std::cout << "df/da: " << std::setprecision(20) << df[a] << '\n';
  std::cout << "df/db: " << std::setprecision(20) << df[b] << '\n';
    // for(auto i: df)
    //     std::cout << std::setprecision(20) << &i.first << '\t' << i.second << '\n';
}